File: 6-TypeNode/backend/database/book-dal.ts
import MysqlPersistence from '../database/mysql-persistence'

type bookDalDTO = {
    id: number | string,
    book: {
        id: number | string,
        title : string,
        author: string,
        cover : string,
        genre : {
            id: number | string,
            name: string
        },
        isbn  : string,
        publication: string
    }
    bdObj: MysqlPersistence,
    conn: typeof globalThis.connection
}

type tableType = {
    bk_id: number | string,
    title : string,
    author: string,
    cover : string,
    gr_id: number | string,
    name: string,
    isbn  : string,
    publication: string
}

const listBooks = async ({ bdObj, conn }: bookDalDTO) => {
    const query = 'SELECT * FROM Book '+
                'inner join Genre on Book.gr_id = Genre.gr_id'
    const res = await bdObj.select({query, conn}) as any
    const books: bookDalDTO["book"][] = []
    res.forEach((book: tableType) => {
        books.push({
            id: book.bk_id,
            title : book.title,
            author: book.author,
            cover : book.cover,
            genre : {
                id: book.gr_id,
                name: book.name
            },
            isbn  : book.isbn,
            publication: book.publication
        })
    });
    return books
}

const read = async ({ id, bdObj, conn }: bookDalDTO) => {
    const query = 'SELECT * FROM Book '+
                'inner join Genre on Book.gr_id = Genre.gr_id ' +
                'where bk_id = ?'
    const book = await bdObj.select({query, conn, param: [id]}) as any
    const res : bookDalDTO["book"] = {
        id: book[0].bk_id,
        title : book[0].title,
        author: book[0].author,
        cover : book[0].cover,
        genre : {
            id: book[0].gr_id,
            name: book[0].name
        },
        isbn  : book[0].isbn,
        publication: book[0].publication
    }
    return res
}

const create = async ({ book, bdObj, conn }: bookDalDTO) => {
    const query = 'insert into Book (title, author, cover, gr_id, isbn, publication) values(?, ?, ?, ?, ?, ?)'
    const res = await bdObj.execute({query, conn, param: [book.title, book.author, book.cover, book.genre.id, book.isbn, book.publication]})
    book.id = res
    return book
}

const update = async ({ book, bdObj, conn }: bookDalDTO) => {
    const query = 'update Book set title = ?, author = ?, cover = ?, '+
                'gr_id = ?, isbn = ?, publication = ? where bk_id = ?'
    const res = await bdObj.execute({query, conn, param: [book.title, book.author, book.cover, book.genre.id, book.isbn, book.publication, book.id]})
    return res
}

const delBook = async ({ id, bdObj, conn }: bookDalDTO) => {
    const res = await bdObj.execute({
        query: 'delete from book where bk_id = ?',
        conn,
        param: [id]
    })
    return res
}

module.exports = {
    listBooks,
    read,
    create,
    update,
    delBook
}

File: 9-Golang/backend/database/book-dal.go
package database

import (
	"database/sql"
	"encoding/json"
)

type Book struct {
	Id          int            `json:"id"`
	Title       string         `json:"title"`
	Author      string         `json:"author"`
	Cover       sql.NullString `json:"-"`
	Genre       Genre          `json:"genre"`
	Isbn        string         `json:"isbn"`
	Publication string         `json:"publication"`
}

func (b Book) MarshalJSON() ([]byte, error) {
	type Alias Book
	return json.Marshal(&struct {
		*Alias
		Cover *string `json:"cover"`
	}{
		Cover: func() *string {
			if b.Cover.Valid {
				return &b.Cover.String
			}
			return nil
		}(),
		Alias: (*Alias)(&b),
	})
}

func GetBooks(conn *Connector) ([]Book, error) {
	books := []Book{}
	rows, err := conn.db.Query(`
		SELECT Book.id, Book.title, Book.author, Book.cover, Book.genre_id, Genre.name AS genre_name, Book.isbn, Book.publication
		FROM Book
		LEFT JOIN Genre ON Book.genre_id = Genre.id
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	for rows.Next() {
		var book Book
		var genreId int
		var genreName string

		err = rows.Scan(&book.Id, &book.Title, &book.Author, &book.Cover, &genreId, &genreName, &book.Isbn, &book.Publication)
		if err != nil {
			return nil, err
		}
		book.Genre = Genre{Id: genreId, Name: genreName}
		books = append(books, book)
	}
	return books, nil
}

func GetBookById(conn *Connector, id int) (Book, error) {
	var book Book
	var genreId int
	var genreName string

	err := conn.db.QueryRow(`
		SELECT Book.id, Book.title, Book.author, Book.cover, Book.genre_id, Genre.name AS genre_name, Book.isbn, Book.publication
		FROM Book
		LEFT JOIN Genre ON Book.genre_id = Genre.id
		WHERE Book.id = ?
	`, id).Scan(&book.Id, &book.Title, &book.Author, &book.Cover, &genreId, &genreName, &book.Isbn, &book.Publication)
	if err != nil {
		return book, err
	}
	book.Genre = Genre{Id: genreId, Name: genreName}

	return book, nil
}

func CreateBook(conn *Connector, book *Book) error {

	_, err := conn.db.Exec("INSERT INTO Book (title, author, genre_id, isbn, publication) VALUES (?, ?, ?, ?, ?)", book.Title, book.Author, book.Genre, book.Isbn, book.Publication)
	if err != nil {
		return err
	}

	return nil
}

func UpdateBook(conn *Connector, book *Book) error {
	_, err := conn.db.Exec("UPDATE Book SET title = ?, author = ?, genre_id = ?, isbn = ?, publication = ? WHERE id = ?", book.Title, book.Author, book.Genre, book.Isbn, book.Publication, book.Id)
	if err != nil {
		return err
	}

	return nil
}

func DeleteBook(conn *Connector, id int) error {
	_, err := conn.db.Exec("DELETE FROM Book WHERE id = ?", id)
	if err != nil {
		return err
	}

	return nil
}

func UpdateBookCover(conn *Connector, oldCover, newCover string) error {
	_, err := conn.db.Exec("UPDATE Book SET cover = ? WHERE cover = ?", newCover, oldCover)
	if err != nil {
		return err
	}

	return nil
}
